cloudinit/sources/__init__.py

```diff
@@ -1149,13 +1150,25 @@ def pkl_load(fname: str) -> Optional[DataSource]:
         return None
 
 
-def parse_cmdline():
+def parse_cmdline() -> str:
     """Check if command line argument for this datasource was passed
     Passing by command line overrides runtime datasource detection
     """
     cmdline = util.get_cmdline()
-    ds_parse_1 = re.search(r"ci\.ds=([a-zA-Z]+)(\s|$)", cmdline)
-    ds_parse_2 = re.search(r"ci\.datasource=([a-zA-Z]+)(\s|$)", cmdline)
-    ds = ds_parse_1 or ds_parse_2
-    if ds:
+    ds_parse_0 = re.search(r"ds=([a-zA-Z]+)(\s|$|;)", cmdline)
+    ds_parse_1 = re.search(r"ci\.ds=([a-zA-Z]+)(\s|$|;)", cmdline)
+    ds_parse_2 = re.search(r"ci\.datasource=([a-zA-Z]+)(\s|$|;)", cmdline)
```

Comment:
All three of these searches should only match with a leading space. We always want to lead with cloud-init cmdline params with our ds config first, and follow with other semi-colon delimited modifiers afterward so we don't have to worry about other applicable configuration characters in front of ds or ci.ds.   Also we don't want to accidentally catch any longer kernelcmdline params that show up in the wild that happen to end with `ds=` such as 
        nfs.callback_nr_threads=
        rcutree.rcu_kick_kthreads= [KNL]

Which would inadvertently cause cloud-init a bit of wasted effort matching those values and finding out that there are no datasourse matching "4" or something.

I think a leading space is safe because every kernel param has to be space-separated so our regex shouldn't expect to extract our leading ds= from a substring of other non-whitespace character strings. Also there are typically other leading params provided before we get to cloud-init extensions like `BOOT_IMAGE` and `root` etc.

because that would affect cloud-init's ability to detect a datasource.
```suggestion
    ds_parse_0 = re.search(r" ds=([a-zA-Z]+)(\s|$|;)", cmdline)
    ds_parse_1 = re.search(r" ci\.ds=([a-zA-Z]+)(\s|$|;)", cmdline)
    ds_parse_2 = re.search(r" ci\.datasource=([a-zA-Z]+)(\s|$|;)", cmdline)
```

Reply:
> I think a leading space is safe because every kernel param has to be space-separated so our regex shouldn't expect to extract our leading ds= from a substring of other non-whitespace character strings. Also there are typically other leading params provided before we get to cloud-init extensions like BOOT_IMAGE and root etc.

Can we guarantee that ds= will never start at the beginning of `/proc/cmdline`? I see that starting the kernel cmdline with `BOOT_IMAGE` is a [grub behavior](https://git.savannah.gnu.org/cgit/grub.git/commit/?id=16ccb8b138218d56875051d547af84410d18f9aa)[1], however, can we rely on this in all cases? It looks like `initrd=` is a [possible argument](https://www.kernel.org/doc/html/v4.14/admin-guide/kernel-parameters.html) on the commandline, so it makes me think that maybe the initramfs possibly doesn't need to be the first argument on the kernel commandline, however, I haven't been able to verify this. I went back and forth about a leading whitespace, but was never able to convince myself that this would never happen (this could be violated by different bootloaders, future changes to grub, etc).

> Also we don't want to accidentally catch any longer kernelcmdline params that show up in the wild that happen to end with ds= such as
> nfs.callback_nr_threads=
> rcutree.rcu_kick_kthreads= [KNL]
>
> Which would inadvertently cause cloud-init a bit of wasted effort matching those . values and finding out that there are no datasourse matching "4" or something.

Agreed, unwanted work is never ideal. However, I made the choice in favor of matching ds=name at the front of the commandline since breaking that case would be a problem, if it is possible. I'd rather get the behavior right in all cases rather than an optimization to avoid string matching in a corner case. If we are confident that this will never happen, I'm fine with making this change, but I'd like to push back in favor of safety here. Thoughts @blackboxsw?

[1] the other args, such as `root=` can be trivially reordered in grub config

Reply:
Works for me. let's leave it as you have it